Igh input validation

globusharris opened the pull request:
Added input validation checks. 

A few notes:
- Covariance has issues marked in it, which I am working on in a separate branch.
- I did a cursory merge of the input validation into the glm statistic but did not bother to do fixes in objective mech that will allow input validation to run (checkRange input validation is run within censorData), since objective mech is not current priority.

Commit:
Refactored checkN

Commit:
Moved input validation functions to seperate file and expanded testing.

Commit:
Added checkNumeric function

Commit:
Expanded functionality of checkEpsilon

Commit:
fixed checkN calls in variance and covariance

Commit:
added checkAccuracy function

Commit:
Expanded checkVariableType and added tests

Commit:
Added numeric check to checkImputationRange and added tests.

Commit:
Expanded checkRange functionality to multiple dimension input

Commit:
Moved censorData test from input validation tests

Commit:
Expanded range validation

Commit:
simplified checkN

Commit:
Simplified checkEpsilon

Commit:
Expanded checkNumeric and checkN to deal with possible NAs and vectors of values & added a mechanism check.

Commit:
Added expectedLength to checkAccuracy

Commit:
Removed excess code that will never get called from checkImputationRange

Commit:
Refactored laplace mech code to integrate checkRange changes and moved laplace utilities to separate file.

Commit:
added laplace utility tests

Commit:
Refactored mean to include input validation.

Commit:
adding to gitignore

Commit:
Refactored histogram to reflect checkRange code changes.

Commit:
Removing .Rhistory file, which should be ignored.

Commit:
Merged input validation into histogram statistic

Commit:
Merged input validation into covariance statistic

Commit:
Merged input validation into tree (though the whole stat is broken so can't test)

Commit:
Merged input validation into variance stat

Commit:
Refactored integration of input validation in mean statistic

Commit:
Cursory merge of input val to glm. Need to update obj mech to support checkRange still. Will do this later.

Commit:
Merged input validation changes to Gaussian mech

Commit:
Merged censorData change due to checkRange changes into all other tests.

Commit:
Added null range tests to mean and variance

Commit:
fixed covariance range validation tests

globusharris commented:
This resolves issues #99, #100, #101

MeganFantes commented on R/mechanism.R:
```diff
@@ -12,7 +12,8 @@
 #' @field delta Differential privacy parameter
-#' @field rng A priori estimate of the variable range
+#' @field rng A priori estimate of the variable range(s).
+#' @field rngFormat One of 'vector' (for single rng input) or 'list' (for multiple rng inputs)
```
Add a couple more words of description here? How many values in the vector?

MeganFantes commented on R/mechanism-objective.R:
```diff
@@ -19,7 +19,7 @@ mechanismObjective$methods(
 
         # censor & impute missing values
-        x <- censorData(x, .self$varType, .self$rng, .self$bins)
+        x <- censorData(x, .self$varType, .self$rng, .self$bins) # NEEDS fixed to support checkRange
```
commenting here just to flag the existing comment about fixing this to support checkRange()

MeganFantes commented on R/mechanism.R:
```diff
@@ -12,7 +12,8 @@
 #' @field delta Differential privacy parameter
-#' @field rng A priori estimate of the variable range
+#' @field rng A priori estimate of the variable range(s).
+#' @field rngFormat One of 'vector' (for single rng input) or 'list' (for multiple rng inputs)
```
Just curious, what is the purpose of this new parameter?

globusharris replied:
It flags if the range is a single range or a sequence of ranges. For backwards compatibility (which, idk, maybe a bad idea), I allow single ranges to be input as a vector, e.g. `c(0,5)`. However, you have to enter multiple ranges as a list (or matrix, though those get automatically coerced to lists) because it's the only way to allow people to enter `NULL` for a single range within the series of ranges, which is how the rest of the library treats empty things (as opposed to using `NA`s rather than `NULL`s). And in the first case, when you pass rng, rng[1]=the minimal bound of the range, and in the second, rng[1] is both the minima and the maxima of the range, so you have to treat them differently.

tldr; if we want to support single ranges entered as vectors as currently done, and support matrices/lists of ranges for more complex statistics and also support some of those ranges potentially having missing values or being completely empty/null, the easiest way to do this is to additionally pass in a parameter that describes how the range is formatted. 

side note: given the way R treats `NULL`, we maybe should actually be using `NA` in most of the places we use `NULL`, but as a computer scientist that feels uncomfortable to me, since most of the time we really do mean `NULL`, it's just that R gets fussy about that. 

It's not the prettiest but it works. There are better ways to do it but I ended up doing this after exploring several options and I think it's ok. Might refactor eventually but given time crunch I think I won't bother unless someone has strong feelings about it

MeganFantes commented on R/statistic-covariance.R:
```diff
@@ -30,7 +30,7 @@ coefficientRelease <- function(formula, release, n) {
 #' @return The sensitivity of the data frame for which the covariance matrix
 #'   is being calculated.
-covarianceSensitivity <- function(n, rng, intercept) {
+covarianceSensitivity <- function(n, rng, intercept) { # THIS IS INCORRECT
```
flagging existing comment

MeganFantes commented on R/statistic-covariance.R:
```diff
@@ -120,25 +120,30 @@ dpCovariance <- setRefClass(
   initialize = function(mechanism, varType, n, epsilon, columns, rng=NULL, imputeRng=NULL, 
-                        intercept=FALSE, formula=NULL, delta=1e-5) {
+                        intercept=FALSE, formula=NULL) {
+    # NOTE THIS IS CURRENTLY NOTE SECURE DUE TO SENS AND EPSILON NOT DISTRIBUTED PROPERLY
```
flagging existing comment about this not being secure

MeganFantes commented on R/statistic-covariance.R:
```diff
@@ -120,25 +120,30 @@ dpCovariance <- setRefClass(
-    .self$delta <- delta
-    .self$rng <- rng
+    .self$mechanism <- checkMechanism(mechanism, "mechanismLaplace")
+    .self$varType <- checkVariableType(varType, c('integer', 'double', 'numeric', 'logical'))  #NEED TO CHANGE TO ALLOW MULTIPLE VARTYPES  
```
flagging existing comment

MeganFantes commented on R/statistic-covariance.R:
```diff
@@ -120,25 +120,30 @@ dpCovariance <- setRefClass(
-    .self$rng <- rng
+    .self$mechanism <- checkMechanism(mechanism, "mechanismLaplace")
+    .self$varType <- checkVariableType(varType, c('integer', 'double', 'numeric', 'logical'))  #NEED TO CHANGE TO ALLOW MULTIPLE VARTYPES  
+    .self$n <- checkN(n)  #NEED TO ADD THING SAYING HOW THIS IS N PER COLUMN
```
could this just be a change where we check n per column within the covariance statistic (e.g. looping through the columns), or would the checkN() function need to be changed?

globusharris replied:
oh, it only actually needs to be one input value for everything since all columns in data frame should be same size (at lease, I've been assuming that, though I suppose we could have that not be true, hmm). my comment was just to remind myself to add a comment in the covariance statistic to make it clear that by n I mean column length not, say, total number of values across all of the columns you want to find covariance of.

MeganFantes commented on R/statistic-covariance.R:
```diff
@@ -120,25 +120,30 @@ dpCovariance <- setRefClass(
+    .self$mechanism <- checkMechanism(mechanism, "mechanismLaplace")
+    .self$varType <- checkVariableType(varType, c('integer', 'double', 'numeric', 'logical'))  #NEED TO CHANGE TO ALLOW MULTIPLE VARTYPES  
+    .self$n <- checkN(n)  #NEED TO ADD THING SAYING HOW THIS IS N PER COLUMN
+    .self$epsilon <- checkEpsilon(epsilon) #NEED TO CHANGE
```
change the function or change the use of the function in the statistic?

globusharris replied:
kinda both, you have to split up global epsilon in covariance statistic. It's not relevant for this PR, relevant to eventual PR on covariance.

MeganFantes commented on R/statistic-covariance.R:
```diff
@@ -120,25 +120,30 @@ dpCovariance <- setRefClass(
+    checkVariableType(typeof(columns), "character")
     
-    if (is.null(imputeRng)) {
+    if (is.null(imputeRng)) { # NEED TO ALLOW FOR IMPUTING ON ONLY SOME OF THE RANGES
```
flagging existing comment

MeganFantes commented on R/statistic-histogram.R:
```diff
@@ -82,10 +82,10 @@ dpHistogram$methods(
         # get the epsilon and accuracy
         if (is.null(epsilon)) {
-            .self$accuracy <- accuracy
+            .self$accuracy <- checkAccuracy(accuracy)
```
why does this checkAccuracy() call not include "expectedLength = 1"?

globusharris replied:
because expectedLength defaults to 1.

MeganFantes commented on R/statistic-histogram.R:
```diff
@@ -82,10 +82,10 @@ dpHistogram$methods(
             .self$epsilon <- histogramGetEpsilon(mechanism, accuracy, delta, alpha, .self$sens)
         } else {
-            .self$epsilon <- epsilon
+            .self$epsilon <- checkEpsilon(epsilon)
```
why does this checkEpsilon() call not include "expectedLength = 1"?

globusharris replied:
because expectedLength defaults to 1.

MeganFantes commented on R/statistic-tree.R:
```diff
@@ -170,22 +170,26 @@ dpTree <- setRefClass(
-        .self$gran <- gran
-        .self$alpha <- alpha
+        .self$n <- checkN(n)
+        .self$rng <- checkRange(rng) # CHANGE
```
flagging existing comment

MeganFantes commented on R/statistic-tree.R:
```diff
@@ -170,22 +170,26 @@ dpTree <- setRefClass(
+        
         if (is.null(epsilon)) {
-            .self$accuracy <- accuracy
+            .self$accuracy <- checkAccuracy(accuracy)
```
why does this not have an "expectedLength = 1" parameter?

globusharris replied:
because expectedLength defaults to 1.

MeganFantes commented on R/statistic-tree.R:
```diff
@@ -170,22 +170,26 @@ dpTree <- setRefClass(
             .self$epsilon <- treeGetParameters(accuracy, rng, gran, alpha)
         } else {
-            .self$epsilon <- epsilon
+            .self$epsilon <- checkEpsilon(epsilon)
```
why does this not have an "expectedLength = 1" parameter?

globusharris replied:
because expectedLength defaults to 1.

MeganFantes commented on R/statistic-variance.R:
```diff
@@ -43,19 +43,24 @@ dpVariance$methods(
+        
         if (is.null(epsilon)) {
-            .self$accuracy <- accuracy
+            .self$accuracy <- checkAccuracy(accuracy)
```
expected length parameter?

globusharris replied:
because expectedLength defaults to 1.

MeganFantes commented on R/statistic-variance.R:
```diff
@@ -43,19 +43,24 @@ dpVariance$methods(
         } else {
             checkEpsilon(epsilon)
-            .self$epsilon <- epsilon
+            .self$epsilon <- checkEpsilon(epsilon)
```
expected length parameter?

globusharris replied:
because expectedLength defaults to 1.

MeganFantes commented on tests/testthat/test-histogram.R:
```diff
@@ -31,19 +31,13 @@ test_that('expect stability mechanism for unknown variable type', {
     expect_error(dpHistogram$new(varType='number data', variable="educ", n=nTest, epsilon=epsilonTest, 
                                  nBins=nBinsTest, delta=deltaTest, rng=c(0,16)), 
-                 "Please enter a data type of 'numeric', 'integer', 'logical', or 'character'")
+                 "Variable type number data should be one of numeric, integer, logical, character", fixed=TRUE)
```
shouldn't the value passed into varType be passed into the error message? Right now the error message seems to be hard coded to have the current value of varType.

globusharris replied:
the error message is hardcoded only in the sense that the expected error in this case is hardcoded in the testthat function. But in general, checkVarType does customize the error message; on a different varType or in a different stat it will output differently.

Commit:
Added docs

Commit:
Merge branch 'develop' of https://github.com/privacytoolsproject/PSI-Library into IGH_inputValidation

globusharris commented:
This resolves issue #99, #100.

ctcovington reviewed (commented):
I'm unable to build this because of an issue in `vignettes/dp-glm.Rmd` (though this is not changed in this PR as far as I can tell). Seems like things work when I delete that file.

globusharris commented:
> I'm unable to build this because of an issue in `vignettes/dp-glm.Rmd` (though this is not changed in this PR as far as I can tell). Seems like things work when I delete that file.

Oh that actually makes sense, because I left out refactoring the objective mechanism with the changes in checkRange, since there are so many other issues with the objective mechanism that it seemed silly to do so. I hadn't thought about how that would result in the vignette not building, since we've mostly been completely ignoring that vignette. I'll push a work-around so that we can keep the vignette (at least for now, it probably shouldn't be included in any release that doesn't also include the obj mech fix).

Commit:
work-around so package builds without glm vignette erroring

Shoeboxam commented on R/statistic-tree.R:
```diff
@@ -170,22 +170,26 @@ dpTree <- setRefClass(
 )
 
 dpTree$methods(
+    # DO NOT USE
```
The initialize function seems like a pretty important function not to use?

globusharris replied:
Correct. This is just marking that the entire tree statistic as currently implemented in develop, does not work. Those fixes are in a separate branch that I haven't pushed yet.

Commit:
Resolved merge conflicts with develop

Merge branch 'develop' of https://github.com/privacytoolsproject/PSI-Library into IGH_inputValidation

# Conflicts:
#	.gitignore
#	R/statistic-covariance.R
#	R/statistic-mean.R
#	R/statistic-variance.R
#	R/utilities-data-validation.R
#	tests/testthat/test-mean.R
#	tests/testthat/test-utilities.R
#	tests/testthat/test-variance.R
